prov/efa/src/rdm/efa_rdm_pke_utils.h

```diff
@@ -94,6 +94,73 @@ efa_rdm_pke_copy_from_hmem_iov(struct efa_mr *iov_mr, struct efa_rdm_pke *pke,
 	return copied;
 }
 
+/** 
+ * @brief This function either posts RDMA read, or sends a NACK packet when p2p
+ * is not available or memory registration limit was reached on the receiver.
+ *
+ * @param[in]    ep         endpoint
+ * @param[in]    pkt_entry  packet entry
+ * @param[in]    rxe        RX entry
+ *
+ * @return 0 on success, or a negative error code.
+ */
+static inline int
+efa_rdm_pke_post_remote_read_or_nack(struct efa_rdm_ep *ep,
+				     struct efa_rdm_pke *pkt_entry,
+				     struct efa_rdm_ope *rxe)
+{
+	int err = 0;
+	int pkt_type;
+
+	pkt_type = efa_rdm_pke_get_base_hdr(pkt_entry)->type;
+
+	if (efa_rdm_ep_use_p2p(ep, rxe->desc[0])) {
+		err = efa_rdm_ope_post_remote_read_or_queue(rxe);
+	} else if (efa_rdm_peer_support_read_nack(rxe->peer)) {
+		EFA_INFO(FI_LOG_EP_CTRL,
+			 "Receiver sending long read "
+			 "NACK packet because P2P is not available, "
+			 "unable to post RDMA read.\n");
+		goto send_nack;
+	} else {
+		EFA_INFO(FI_LOG_EP_CTRL, "P2P is not available, "
+					 "unable to post RDMA read.\n");
+		err = -FI_EOPNOTSUPP;
+	}
+
+	if (err == -FI_ENOMR) {
+		if (efa_rdm_peer_support_read_nack(rxe->peer)) {
+			EFA_INFO(FI_LOG_EP_CTRL, "Receiver sending long read "
+						 "NACK packet because memory "
+						 "registration limit was "
+						 "reached on the receiver.\n");
+			goto send_nack;
+		} else {
+			/* Peer does not support the READ_NACK packet. So we
+			 * return EAGAIN and hope that the app runs progress
+			 * again which will free some MR registrations */
+			err = -FI_EAGAIN;
+		}
+	}
+
+	return err;
+
+send_nack:
+	rxe->internal_flags |= EFA_RDM_OPE_READ_NACK;
+	/* Only set the flag for runting read. The NACK
+	 * packet is sent after all runting read
+	 * RTM packets have been received */
+	if (!efa_rdm_pkt_type_is_runtread(pkt_type)) {
```

Comment:
`efa_rdm_rxe_map_insert` is a protocoll implementation detail, and only works for pkt type that has msg_id. So we shouldn't check the op code here, but check the pkt type. Since msg_id is only present in rtm pkt type, it should be something like `efa_rdm_pkt_type_is_rtm`

For cleaner logic, can we also refactor the code block to 

```
if (efa_rdm_pkt_type_is_runtread(pkt_type)) /* do nothing for runting read */
   return 0;

if (efa_rdm_pkt_type_is_rtm(pkt_type))
   efa_rdm_rxe_map_insert(&ep->rxe_map, pkt_entry, rxe);

err = efa_rdm_ope_post_send_or_queue(rxe, EFA_RDM_READ_NACK_PKT);
```